src/utils/math.hpp

```diff
@@ -249,32 +143,19 @@ constexpr unsigned int count_leading_zeros_impl(N n, std::size_t w) {
  * @see count_leading_ones()
  */
 template<typename N>
-constexpr unsigned int count_leading_zeros(N n) {
-#if defined(__GNUC__) || defined(__clang__)
-	// GCC’s `__builtin_clz` returns an undefined value when called with 0
-	// as argument.
-	// [<http://gcc.gnu.org/onlinedocs/gcc/Other-Builtins.html>]
-	if (n == 0) {
-		// Return the quantity of zero bits in `n` rather than
-		// returning that undefined value.
-		return static_cast<unsigned int>(bit_width(n));
+constexpr std::enable_if_t<std::is_integral_v<N>, unsigned int> count_leading_zeros(N n) {
+	const auto x = static_cast<std::make_unsigned_t<N>>(n);
+	constexpr decltype(x) mask{1};
```

Comment:
You don't need to flag this constexpr. It's already inside a constexpr function. I'm not actually even sure what this would do...?

Reply:
Making it constexpr says that I (as author) think it can be fully evaluated at compile time. I think there's no practical difference between constexpr and const here; the compiler is likely to optimise it away anyway.

That the containing function is constexpr doesn't affect this - `count_leading_zeros(n)` can only be evaluated at compile time if `n` is itself a compile time constant expression. For the callers that are handing utf8 data at runtime, the containing function might as well not be constexpr at all, IIUC.

Reply:
Right, but in the compile-time case it doesn't make a difference.